README on how to use the template

4n4nd opened the pull request.

4n4nd commented:
Oops my bad, I should have created a new PR for the last commit (Update README).

durandom commented on openshift/mlflow-experiment-image-buildconfig.yaml:
```diff
@@ -0,0 +1,69 @@
+apiVersion: v1
```
remove `-image-` from all the naming. A build config always produces an `image` :)

durandom commented on openshift/mlflow-experiment-image-buildconfig.yaml:
```diff
@@ -0,0 +1,69 @@
+  name: APPLICATION_NAME
+  required: true
+- name: GIT_URI
+  value: https://github.com/4n4nd/experiment-tracking-template.git
```
default URL should point to https://github.com/AICoE/experiment-tracking-template.git

durandom commented on openshift/mlflow-experiment-job-template.yaml:
```diff
@@ -0,0 +1,52 @@
+  name: APPLICATION_NAME
+  required: true
+- name: MLFLOW_TRACKING_URI
+  value: http://mlflow-server-test-route-aiops-kubeflow.cloud.paas.upshift.redhat.com/
```
if we're running the experiment in the same namespace as the mlflow server, I think we can use internal routing. 
`mlflow-server-test-service.aiops-kubeflow.svc`
Maybe it even works without the `aiops-kubeflow.svc` suffix.

Or look at these environment variables, which are defined for every pod
```
LFLOW_SERVER_SVC_PORT=tcp://172.56.132.180:5000
MLFLOW_SERVER_SVC_PORT_5000_TCP=tcp://172.56.132.180:5000
MLFLOW_SERVER_SVC_PORT_5000_TCP_ADDR=172.56.132.180
MLFLOW_SERVER_SVC_PORT_5000_TCP_PORT=5000
MLFLOW_SERVER_SVC_PORT_5000_TCP_PROTO=tcp
MLFLOW_SERVER_SVC_SERVICE_HOST=172.56.132.180
MLFLOW_SERVER_SVC_SERVICE_PORT=5000
MLFLOW_SERVER_SVC_SERVICE_PORT_HTTP=5000
MLFLOW_SERVER_TEST_SERVICE_PORT=tcp://172.56.45.193:5000
MLFLOW_SERVER_TEST_SERVICE_PORT_5000_TCP=tcp://172.56.45.193:5000
MLFLOW_SERVER_TEST_SERVICE_PORT_5000_TCP_ADDR=172.56.45.193
MLFLOW_SERVER_TEST_SERVICE_PORT_5000_TCP_PORT=5000
MLFLOW_SERVER_TEST_SERVICE_PORT_5000_TCP_PROTO=tcp
MLFLOW_SERVER_TEST_SERVICE_SERVICE_HOST=172.56.45.193
MLFLOW_SERVER_TEST_SERVICE_SERVICE_PORT=5000
MLFLOW_SERVER_TEST_SERVICE_SERVICE_PORT_HTTP=5000
```

maybe we default to one of those, if `MLFLOW_TRACKING_URI` is not set, i.e. leave it blank here

zak-hassan replied:
My recommendation is to make it required for the template to have the MLFLOW_TRACKING_URI set.

durandom replied:
but if we're running in the same namespace as the tracking server and we could infer the url of the tracking server that would be one parameter less. 
if we can't make it easily, then so be it...

zak-hassan replied:
Should be fine. But I'm saying mlflow will look for that environment variable to be set so either way we will need to set that env var to have it communicate with the tracking server.

4n4nd replied:
I will change it  to `MLFLOW_TRACKING_URI=http://$(MLFLOW_SERVER_TEST_SERVICE_SERVICE_HOST):$(MLFLOW_SERVER_TEST_SERVICE_SERVICE_PORT)` so it uses the internal route

durandom commented on openshift/mlflow-experiment-job-template.yaml:
```diff
@@ -0,0 +1,52 @@
+              memory: 1Gi
+          env:
+          - name: PARAM_ALPHA     # Using env variables to pass parameters
+            value: "0.1"
```
these `PARAMS` must not be in the template. They are being set when creating a job from the template

zak-hassan replied:
+1 We should remove the PARAM_ALPHA, PARAM_L1_RATION as this is a custom parameter passed in to the mlflow sample code.

4n4nd replied:
They were supposed to be examples, but since DSs are not going to use this template, I will remove them

durandom reviewed (changes requested):
the README can be in this PR, that's ok.
But the templates should go into https://github.com/AICoE/experiment-tracking because they are created by the project admin. Then the datascientist (DS) actually creates a BC from the template and a job from the job template.

This repo here is just the boilerplate for the experiment

durandom commented:
@zmhassan would you also please review?

durandom commented on README.md:
```diff
@@ -1,3 +1,78 @@
+# Tracking MLflow jobs on Openshift
 
-Demo of tracking machine learning metrics into mlflow running on k8s
+This is a demo application which includes templates that can be used to run mlflow jobs on Openshift.
```
scratch the templates part. these are created from the other repo

durandom commented on README.md:
```diff
@@ -1,3 +1,78 @@
+This is a demo application which includes templates that can be used to run mlflow jobs on Openshift.
+
+## Running on OpenShift
+This is a two stage process, because you need to create the image only once and then you can create multiple containers using the same image but with different parameters.
```
I wouldnt even mention images and container. talk about `the code` and `experiment`.

4n4nd replied:
@durandom I tried to simplify this a little more, so if you could give it another read

durandom commented on README.md:
````diff
@@ -1,3 +1,78 @@
+To initiate a new image build, run the following command from the root of this repository:
+
+```
+oc new-app ./openshift/mlflow-experiment-image-buildconfig.yaml --param APPLICATION_NAME=mlflow-experiment-image-demo --param GIT_URI=https://github.com/AICoE/experiment-tracking-template.git --env APP_FILE=app.py
````
so the flow would be 

1. admin creates template in openshift via `oc create template.yaml`
2. DS uses template to create his BC via `oc process mlflow-experiment-bc | oc create -f -`

I think `oc process` is better here, because `new-app` does way more than we need

zak-hassan replied:
@4n4nd You can use this sample mlflow project to demonstrate building the image. Also note that there are a few environment variables you want to set:

* `MLFLOW_TRACKING_URI`
* `MLFLOW_EXPERIMENT_ID`

zak-hassan replied:
Try to use this https://github.com/zmhassan/sample-mlflow-tracking-demo git repo to demo it running.

durandom replied:
this repo here is a fork of yours :)
and the MLFLOW_TRACKING_URI and MLFLOW_EXPERIMENT_ID should be defaulted. So that the bar to run something is reaaaally low

4n4nd replied:
Instead of MLFLOW_EXPERIMENT_ID, I am planning to use just MLFLOW_EXPERIMENT_NAME which should be easier to use for DSs

durandom commented on README.md:
````diff
@@ -1,3 +1,78 @@
+
+From the root of this repository, run the following command:
+```
+oc new-app ./openshift/mlflow-experiment-job-template.yaml --APPLICATION_NAME=mlflow-experiment-job-1 --param APP_IMAGE_NAME=mlflow-experiment-image-demo --param MLFLOW_TRACKING_URI=http://mlflow-server-url/ --env PARAM_ALPHA=0.4 --env PARAM_L1_RATIO=0.8
````
same here, I guess `oc process` might be better. But I might be wrong...

4n4nd replied:
you cannot specify --env in `oc process` so I might switch back to `oc new-app`

durandom commented on README.md:
```diff
@@ -1,3 +1,78 @@
+<br/>You can set `GIT_URI` to the link for your repository to create a container image for your application.
+<br/>You can also change the `APP_FILE` env variable to your app file name.
+
+Bulding the container image can take a couple of minutes or more depending on the number of packages that need to be installed from the dependencies list.
```
please add a line on how to trigger a new build. `oc start-build ...`

zak-hassan commented on openshift/mlflow-experiment-job-template.yaml:
```diff
@@ -0,0 +1,52 @@
+  required: true
+- name: MLFLOW_TRACKING_URI
+  value: http://mlflow-server-test-route-aiops-kubeflow.cloud.paas.upshift.redhat.com/
+  required: false
```
We should make this `required: true` because we want the user of mlflow to specify their tracking server address.

4n4nd replied:
I will change it to required and change its default value

zak-hassan commented on openshift/mlflow-experiment-image-buildconfig.yaml:
```diff
@@ -0,0 +1,69 @@
+    output:
+      to:
+        kind: ImageStreamTag
+        name: ${APPLICATION_NAME}:latest
```
Let's remove `:latest` and as users may want to use different tags

4n4nd replied:
@zmhassan if not `:latest`, what should I use as the tag?

zak-hassan replied:
Leave it blank. Perhaps have another parameter called `IMAGE_URL`.

4n4nd replied:
@zmhassan In the build-config template? or the job template?

zak-hassan replied:
So the purpose of the build-config i thought was to build the image. Then once you have a ready image you could run it in a job template with different parameters.

4n4nd replied:
I will make the image being used in the job-template configurable then

zak-hassan reviewed (approved):
Its okay to include that. Perhaps your testing `mlflow run` or `mlflow server` command in this project. Perhaps thats why you include `mlruns` in your git ignore.

zak-hassan reviewed (changes requested):
I'm going to wait till all requested changes are resolved and then approve this.

4n4nd commented:
Once the templates are checked here, I will remove them and push them to the experiment-tracking repo [Here](https://github.com/AICoE/experiment-tracking)

durandom commented on README.md:
```diff
@@ -1,23 +1,30 @@
-### Create a container image
-To run this application on openshift, we need to create a container image for it.  To do this we have a image build template [here in the repository](https://github.com/AICoE/experiment-tracking-template/blob/master/openshift/mlflow-experiment-image-buildconfig.yaml).
+* The first stage is to get your code on openshift
+* The second stage is to run the experiment.<br/>
```
no `<br/>` tag

durandom commented on README.md:
````diff
@@ -1,23 +1,30 @@
 
 ```
-oc new-app ./openshift/mlflow-experiment-image-buildconfig.yaml --param APPLICATION_NAME=mlflow-experiment-image-demo --param GIT_URI=https://github.com/AICoE/experiment-tracking-template.git --env APP_FILE=app.py
+oc process mlflow-experiment-bc --param APPLICATION_NAME=mlflow-experiment-1-image --param GIT_URI=https://github.com/AICoE/experiment-tracking-template.git --env APP_FILE=app.py | oc create -f -
````
remove `-1-image` suffix from APPLICATION_NAME. This creates the BC, that will never change, so no need to add numbers to the name.
Maybe call it `my-mlflow-experiment` to indicate that the DS should pick his name

4n4nd replied:
good idea

durandom commented on README.md:
```diff
@@ -1,23 +1,30 @@
-Here, the `APPLICATION_NAME` parameter is the name of the container image that will be built.
-<br/>You can set `GIT_URI` to the link for your repository to create a container image for your application.
-<br/>You can also change the `APP_FILE` env variable to your app file name.
+Here, the `APPLICATION_NAME` parameter is the name of the container image that will be built and is where your code will be downloaded, this is the image name that you will use to run your experiments.<br/>
```
no `<br/`> New paragraph in markown are just double blank lines - a new line is one single blank line

4n4nd replied:
The rest line-breaks were intentional, but I think I will just change the lines to bullet points instead

durandom commented on README.md:
```diff
@@ -1,23 +1,30 @@
-<br/>You can set `GIT_URI` to the link for your repository to create a container image for your application.
-<br/>You can also change the `APP_FILE` env variable to your app file name.
+Here, the `APPLICATION_NAME` parameter is the name of the container image that will be built and is where your code will be downloaded, this is the image name that you will use to run your experiments.<br/>
+You can set `GIT_URI` to the link for your repository to create a container image for your application.<br/>
```
same `</br>`

durandom commented on README.md:
````diff
@@ -34,21 +41,41 @@ buildconfig.build.openshift.io "mlflow-experiment-image-demo" created
+To run the experiment, use the following command:
 ```
-oc new-app ./openshift/mlflow-experiment-job-template.yaml --APPLICATION_NAME=mlflow-experiment-job-1 --param APP_IMAGE_NAME=mlflow-experiment-image-demo --param MLFLOW_TRACKING_URI=http://mlflow-server-url/ --env PARAM_ALPHA=0.4 --env PARAM_L1_RATIO=0.8
+oc process mlflow-experiment-job --param APPLICATION_NAME=mlflow-experiment-1-run-1 --param APP_IMAGE_URI=mlflow-experiment-1-image --env PARAM_ALPHA=0.4 --env PARAM_L1_RATIO=0.8 | oc create -f -
````
can the `APPLICATION_NAME` be auto generated here?

4n4nd replied:
I can create a new parameter `RUN_NAME` and set `APPLICATION_NAME` something generated from `RUN_NAME`
UPDATE: The above suggestion did not work

durandom commented on README.md:
````diff
@@ -34,21 +41,41 @@ buildconfig.build.openshift.io "mlflow-experiment-image-demo" created
 ```
-In the above command, `MLFLOW_TRACKING_URI` is the url for your mlflow tracking server, `PARAM_ALPHA` and `PARAM_L1_RATIO` are the parameters for training the demo model.
-<br/>Make sure that the `APP_IMAGE_NAME` that you specified in the above command is the same as the one that you built in the previous stage.
+In the above command, `PARAM_ALPHA` and `PARAM_L1_RATIO` are the parameters for training the demo model. ([Here](https://github.com/AICoE/experiment-tracking-template/blob/master/app.py#L90-L91))<br/>
````
if linking to something on github, especially lines, always use the sha link. You get that by pressing `y` on the source code page. because the `app.py` will change and then you're linking to something else

in this case: `https://github.com/AICoE/experiment-tracking-template/blob/23efac85099ca40b1bc2ead402008a8febf69608/app.py#L90-L91`

durandom reviewed (changes requested):
great progress on the documentation side ❤️

durandom reviewed (changes requested):
great stuff, thanks

one more thing ™️ 
* remove the templates from this PR
* squash the commits into one

Commit:
Update README.md
Fix app.py
Update Pipfile.lock
Add mlruns/ to .gitignore to prevent run artifacts from being pushed
